File: dogreal_app/lib/models/post.dart
class PostImage {
  final String url;
  final String type; // 'front' or 'back'

  PostImage({
    required this.url,
    required this.type,
  });

  factory PostImage.fromJson(Map<String, dynamic> json) {
    return PostImage(
      url: json['url'] ?? '',
      type: json['type'] ?? 'front',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'url': url,
      'type': type,
    };
  }
}

class Comment {
  final String id;
  final String userId;
  final String username;
  final String text;
  final DateTime createdAt;

  Comment({
    required this.id,
    required this.userId,
    required this.username,
    required this.text,
    required this.createdAt,
  });

  factory Comment.fromJson(Map<String, dynamic> json) {
    return Comment(
      id: json['_id'] ?? '',
      userId: json['user'] is String ? json['user'] : (json['user']?['_id'] ?? ''),
      username: json['user'] is String ? '' : (json['user']?['username'] ?? ''),
      text: json['text'] ?? '',
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}

class Post {
  final String id;
  final String userId;
  final String username;
  final String petId;
  final String petName;
  final String petBreed;
  final String petAvatar;
  final int petAge;
  final List<PostImage> images;
  final String caption;
  final String location;
  final List<String> likes;
  final int likesCount;
  final List<Comment> comments;
  final int commentsCount;
  final bool isLate;
  final DateTime postedAt;
  final DateTime createdAt;
  final bool isActive;

  Post({
    required this.id,
    required this.userId,
    required this.username,
    required this.petId,
    required this.petName,
    required this.petBreed,
    required this.petAvatar,
    required this.petAge,
    required this.images,
    required this.caption,
    required this.location,
    required this.likes,
    required this.likesCount,
    required this.comments,
    required this.commentsCount,
    required this.isLate,
    required this.postedAt,
    required this.createdAt,
    required this.isActive,
  });

  factory Post.fromJson(Map<String, dynamic> json) {
    // 处理用户信息
    final user = json['user'];
    final userId = user is String ? user : (user?['_id'] ?? '');
    final username = user is String ? '' : (user?['username'] ?? '');

    // 处理宠物信息
    final pet = json['pet'];
    final petId = pet is String ? pet : (pet?['_id'] ?? '');
    final petName = pet is String ? '' : (pet?['name'] ?? '');
    final petBreed = pet is String ? '' : (pet?['breed'] ?? '');
    final petAvatar = pet is String ? '' : (pet?['avatar'] ?? '');
    final petAge = pet is String ? 0 : (pet?['age'] ?? 0);

    // 处理图片
    final imagesList = (json['images'] as List?)
        ?.map((img) => PostImage.fromJson(img))
        .toList() ?? [];

    // 处理点赞
    final likesList = (json['likes'] as List?)
        ?.map((like) => like.toString())
        .toList() ?? [];

    // 处理评论
    final commentsList = (json['comments'] as List?)
        ?.map((comment) => Comment.fromJson(comment))
        .toList() ?? [];

    return Post(
      id: json['_id'] ?? '',
      userId: userId,
      username: username,
      petId: petId,
      petName: petName,
      petBreed: petBreed,
      petAvatar: petAvatar,
      petAge: petAge,
      images: imagesList,
      caption: json['caption'] ?? '',
      location: json['location'] ?? '',
      likes: likesList,
      likesCount: json['likesCount'] ?? 0,
      comments: commentsList,
      commentsCount: json['commentsCount'] ?? 0,
      isLate: json['isLate'] ?? false,
      postedAt: DateTime.parse(json['postedAt'] ?? json['createdAt']),
      createdAt: DateTime.parse(json['createdAt']),
      isActive: json['isActive'] ?? true,
    );
  }

  Post copyWith({
    int? likesCount,
    List<String>? likes,
    int? commentsCount,
    List<Comment>? comments,
  }) {
    return Post(
      id: id,
      userId: userId,
      username: username,
      petId: petId,
      petName: petName,
      petBreed: petBreed,
      petAvatar: petAvatar,
      petAge: petAge,
      images: images,
      caption: caption,
      location: location,
      likes: likes ?? this.likes,
      likesCount: likesCount ?? this.likesCount,
      comments: comments ?? this.comments,
      commentsCount: commentsCount ?? this.commentsCount,
      isLate: isLate,
      postedAt: postedAt,
      createdAt: createdAt,
      isActive: isActive,
    );
  }
}

File: backend/src/models/Post.js
const mongoose = require('mongoose');

const postSchema = new mongoose.Schema({
  user: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  pet: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Pet',
    required: true
  },
  images: [{
    url: {
      type: String,
      required: true
    },
    type: {
      type: String,
      enum: ['front', 'back'],
      required: true
    }
  }],
  caption: {
    type: String,
    maxlength: [500, '描述不能超过500个字符'],
    default: ''
  },
  location: {
    type: String,
    default: ''
  },
  likes: [{
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User'
  }],
  likesCount: {
    type: Number,
    default: 0
  },
  comments: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: true
    },
    text: {
      type: String,
      required: true,
      maxlength: 200
    },
    createdAt: {
      type: Date,
      default: Date.now
    }
  }],
  commentsCount: {
    type: Number,
    default: 0
  },
  isLate: {
    type: Boolean,
    default: false
  },
  postedAt: {
    type: Date,
    default: Date.now
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// 索引优化
postSchema.index({ user: 1, createdAt: -1 });
postSchema.index({ pet: 1, createdAt: -1 });
postSchema.index({ postedAt: -1 });
postSchema.index({ isActive: 1, createdAt: -1 });

// 虚拟字段：检查是否是今天发布的
postSchema.virtual('isToday').get(function() {
  const now = new Date();
  const posted = new Date(this.postedAt);
  return (
    now.getFullYear() === posted.getFullYear() &&
    now.getMonth() === posted.getMonth() &&
    now.getDate() === posted.getDate()
  );
});

// 更新点赞数
postSchema.methods.updateLikesCount = function() {
  this.likesCount = this.likes.length;
  return this.save();
};

// 更新评论数
postSchema.methods.updateCommentsCount = function() {
  this.commentsCount = this.comments.length;
  return this.save();
};

// 检查用户是否点赞
postSchema.methods.isLikedBy = function(userId) {
  return this.likes.some(like => like.toString() === userId.toString());
};

module.exports = mongoose.model('Post', postSchema);
